// @pakor/wacap-wrapper/dist/types/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WacapEventType = void 0;
/**
 * Event types for webhook support
 */
var WacapEventType;
(function (WacapEventType) {
    // Connection events
    WacapEventType["CONNECTION_UPDATE"] = "connection.update";
    WacapEventType["CONNECTION_OPEN"] = "connection.open";
    WacapEventType["CONNECTION_CLOSE"] = "connection.close";
    WacapEventType["QR_CODE"] = "qr.code";
    // Message events
    WacapEventType["MESSAGE_RECEIVED"] = "message.received";
    WacapEventType["MESSAGE_SENT"] = "message.sent";
    WacapEventType["MESSAGE_UPDATE"] = "message.update";
    WacapEventType["MESSAGE_DELETE"] = "message.delete";
    // Chat events
    WacapEventType["CHAT_UPDATE"] = "chat.update";
    WacapEventType["CHAT_DELETE"] = "chat.delete";
    // Contact events
    WacapEventType["CONTACT_UPDATE"] = "contact.update";
    // Group events
    WacapEventType["GROUP_UPDATE"] = "group.update";
    WacapEventType["GROUP_PARTICIPANTS_UPDATE"] = "group.participants.update";
    // Presence events
    WacapEventType["PRESENCE_UPDATE"] = "presence.update";
    // Call events
    WacapEventType["CALL"] = "call";
    // Session events
    WacapEventType["SESSION_START"] = "session.start";
    WacapEventType["SESSION_STOP"] = "session.stop";
    WacapEventType["SESSION_ERROR"] = "session.error";
})(WacapEventType || (exports.WacapEventType = WacapEventType = {}));
