// @pakor/wacap-wrapper/dist/storage/sqlite-adapter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SQLiteStorageAdapter = void 0;
const better_sqlite3_1 = __importDefault(require("better-sqlite3"));
const path_1 = require("path");
const fs_1 = require("fs");
/**
 * SQLite storage adapter for session data
 */
class SQLiteStorageAdapter {
    db;
    dbPath;
    constructor(baseDir = './sessions') {
        // Ensure sessions directory exists
        if (!(0, fs_1.existsSync)(baseDir)) {
            (0, fs_1.mkdirSync)(baseDir, { recursive: true });
        }
        this.dbPath = (0, path_1.join)(baseDir, 'wacap.db');
        this.db = new better_sqlite3_1.default(this.dbPath);
    }
    async init() {
        // Create sessions table
        this.db.exec(`
      CREATE TABLE IF NOT EXISTS sessions (
        session_id TEXT PRIMARY KEY,
        creds TEXT NOT NULL,
        created_at INTEGER NOT NULL,
        updated_at INTEGER NOT NULL
      )
    `);
        // Create messages table
        this.db.exec(`
      CREATE TABLE IF NOT EXISTS messages (
        id TEXT PRIMARY KEY,
        session_id TEXT NOT NULL,
        jid TEXT NOT NULL,
        message TEXT NOT NULL,
        timestamp INTEGER NOT NULL,
        from_me INTEGER NOT NULL,
        FOREIGN KEY (session_id) REFERENCES sessions(session_id) ON DELETE CASCADE
      )
    `);
        // Create contacts table
        this.db.exec(`
      CREATE TABLE IF NOT EXISTS contacts (
        id TEXT PRIMARY KEY,
        session_id TEXT NOT NULL,
        jid TEXT NOT NULL,
        name TEXT,
        notify TEXT,
        data TEXT NOT NULL,
        UNIQUE(session_id, jid),
        FOREIGN KEY (session_id) REFERENCES sessions(session_id) ON DELETE CASCADE
      )
    `);
        // Create chats table
        this.db.exec(`
      CREATE TABLE IF NOT EXISTS chats (
        id TEXT PRIMARY KEY,
        session_id TEXT NOT NULL,
        jid TEXT NOT NULL,
        data TEXT NOT NULL,
        timestamp INTEGER NOT NULL,
        UNIQUE(session_id, jid),
        FOREIGN KEY (session_id) REFERENCES sessions(session_id) ON DELETE CASCADE
      )
    `);
        // Create indexes for better performance
        this.db.exec(`
      CREATE INDEX IF NOT EXISTS idx_messages_session_jid ON messages(session_id, jid);
      CREATE INDEX IF NOT EXISTS idx_contacts_session ON contacts(session_id);
      CREATE INDEX IF NOT EXISTS idx_chats_session ON chats(session_id);
    `);
    }
    async saveSession(sessionId, creds) {
        const now = Date.now();
        const stmt = this.db.prepare(`
      INSERT INTO sessions (session_id, creds, created_at, updated_at)
      VALUES (?, ?, ?, ?)
      ON CONFLICT(session_id) DO UPDATE SET
        creds = excluded.creds,
        updated_at = excluded.updated_at
    `);
        stmt.run(sessionId, JSON.stringify(creds), now, now);
    }
    async loadSession(sessionId) {
        const stmt = this.db.prepare(`
      SELECT creds FROM sessions WHERE session_id = ?
    `);
        const row = stmt.get(sessionId);
        return row ? JSON.parse(row.creds) : null;
    }
    async deleteSession(sessionId) {
        const stmt = this.db.prepare(`
      DELETE FROM sessions WHERE session_id = ?
    `);
        stmt.run(sessionId);
    }
    async hasSession(sessionId) {
        const stmt = this.db.prepare(`
      SELECT 1 FROM sessions WHERE session_id = ? LIMIT 1
    `);
        return stmt.get(sessionId) !== undefined;
    }
    async saveMessage(sessionId, message) {
        this.ensureSessionExists(sessionId);
        const messageId = String(message.key.id || "");
        const jid = String(message.key.remoteJid || "");
        // Convert timestamp Long → Number
        const rawTs = message.messageTimestamp;
        const timestamp = typeof rawTs === "object" && rawTs?.toNumber
            ? rawTs.toNumber()
            : Number(rawTs) || Date.now();
        const fromMe = message.key.fromMe ? 1 : 0;
        const stmt = this.db.prepare(`
        INSERT INTO messages (id, session_id, jid, message, timestamp, from_me)
        VALUES (?, ?, ?, ?, ?, ?)
        ON CONFLICT(id) DO UPDATE SET
        message = excluded.message,
        timestamp = excluded.timestamp
    `);
        stmt.run(messageId, sessionId, jid, JSON.stringify(message), timestamp, fromMe);
    }
    async getMessages(sessionId, jid, limit = 50) {
        const stmt = this.db.prepare(`
      SELECT message FROM messages
      WHERE session_id = ? AND jid = ?
      ORDER BY timestamp DESC
      LIMIT ?
    `);
        const rows = stmt.all(sessionId, jid, limit);
        return rows.map(row => JSON.parse(row.message));
    }
    async saveContact(sessionId, contact) {
        this.ensureSessionExists(sessionId);
        const stmt = this.db.prepare(`
      INSERT INTO contacts (id, session_id, jid, name, notify, data)
      VALUES (?, ?, ?, ?, ?, ?)
      ON CONFLICT(session_id, jid) DO UPDATE SET
        name = excluded.name,
        notify = excluded.notify,
        data = excluded.data
    `);
        const id = `${sessionId}:${contact.id}`;
        stmt.run(id, sessionId, contact.id, contact.name || null, contact.notify || null, JSON.stringify(contact));
    }
    async getContacts(sessionId) {
        const stmt = this.db.prepare(`
      SELECT data FROM contacts WHERE session_id = ?
    `);
        const rows = stmt.all(sessionId);
        return rows.map(row => JSON.parse(row.data));
    }
    async saveChat(sessionId, chat) {
        this.ensureSessionExists(sessionId);
        const jid = chat.id;
        const timestamp = Date.now();
        const stmt = this.db.prepare(`
      INSERT INTO chats (id, session_id, jid, data, timestamp)
      VALUES (?, ?, ?, ?, ?)
      ON CONFLICT(session_id, jid) DO UPDATE SET
        data = excluded.data,
        timestamp = excluded.timestamp
    `);
        const id = `${sessionId}:${jid}`;
        stmt.run(id, sessionId, jid, JSON.stringify(chat), timestamp);
    }
    async getChats(sessionId) {
        const stmt = this.db.prepare(`
      SELECT data FROM chats WHERE session_id = ? ORDER BY timestamp DESC
    `);
        const rows = stmt.all(sessionId);
        return rows.map(row => JSON.parse(row.data));
    }
    async close() {
        this.db.close();
    }
    /**
     * Ensure session record exists before saving related data
     */
    ensureSessionExists(sessionId) {
        const stmt = this.db.prepare(`
      INSERT OR IGNORE INTO sessions (session_id, creds, created_at, updated_at)
      VALUES (?, ?, ?, ?)
    `);
        const now = Date.now();
        stmt.run(sessionId, '{}', now, now);
    }
    /**
     * Get database instance for advanced usage
     */
    getDatabase() {
        return this.db;
    }
}
exports.SQLiteStorageAdapter = SQLiteStorageAdapter;
