// @pakor/wacap-wrapper/dist/events/event-manager.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventManager = void 0;
const eventemitter3_1 = __importDefault(require("eventemitter3"));
const types_1 = require("../types");
/**
 * Event manager for handling WhatsApp events
 * Provides a simple interface for webhook integration
 */
class EventManager {
    emitter;
    sessionId;
    constructor(sessionId) {
        this.emitter = new eventemitter3_1.default();
        this.sessionId = sessionId;
    }
    /**
     * Register an event handler
     */
    on(event, handler) {
        this.emitter.on(event, handler);
    }
    /**
     * Register a one-time event handler
     */
    once(event, handler) {
        this.emitter.once(event, handler);
    }
    /**
     * Remove an event handler
     */
    off(event, handler) {
        this.emitter.off(event, handler);
    }
    /**
     * Remove all event handlers
     */
    removeAllListeners(event) {
        this.emitter.removeAllListeners(event);
    }
    /**
     * Emit an event
     */
    emit(event, data) {
        this.emitter.emit(event, data);
    }
    /**
     * Convenience method: Listen for connection updates
     */
    onConnectionUpdate(handler) {
        this.on(types_1.WacapEventType.CONNECTION_UPDATE, handler);
    }
    /**
     * Convenience method: Listen for connection open
     */
    onConnectionOpen(handler) {
        this.on(types_1.WacapEventType.CONNECTION_OPEN, handler);
    }
    /**
     * Convenience method: Listen for connection close
     */
    onConnectionClose(handler) {
        this.on(types_1.WacapEventType.CONNECTION_CLOSE, handler);
    }
    /**
     * Convenience method: Listen for QR code
     */
    onQRCode(handler) {
        this.on(types_1.WacapEventType.QR_CODE, handler);
    }
    /**
     * Convenience method: Listen for received messages
     */
    onMessageReceived(handler) {
        this.on(types_1.WacapEventType.MESSAGE_RECEIVED, handler);
    }
    /**
     * Convenience method: Listen for sent messages
     */
    onMessageSent(handler) {
        this.on(types_1.WacapEventType.MESSAGE_SENT, handler);
    }
    /**
     * Convenience method: Listen for message updates
     */
    onMessageUpdate(handler) {
        this.on(types_1.WacapEventType.MESSAGE_UPDATE, handler);
    }
    /**
     * Convenience method: Listen for group participant updates
     */
    onGroupParticipantsUpdate(handler) {
        this.on(types_1.WacapEventType.GROUP_PARTICIPANTS_UPDATE, handler);
    }
    /**
     * Convenience method: Listen for presence updates
     */
    onPresenceUpdate(handler) {
        this.on(types_1.WacapEventType.PRESENCE_UPDATE, handler);
    }
    /**
     * Get the underlying EventEmitter for advanced usage
     */
    getEmitter() {
        return this.emitter;
    }
}
exports.EventManager = EventManager;
